WordPress/Classes/ViewRelated/System/WPTabBarController.m

```diff
@@ -435,4 +454,108 @@ - (void)defaultAccountDidChange:(NSNotification *)notification
+
+- (void)addNotificationBadgeIcon
+{
+    self.notificationBadgeIconView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, WPNotificationBadgeIconSize, WPNotificationBadgeIconSize)];
+    self.notificationBadgeIconView.image = [UIImage imageWithColor:[WPStyleGuide jazzyOrange]
+                                                        havingSize:CGSizeMake(WPNotificationBadgeIconSize, WPNotificationBadgeIconSize)];
+    self.notificationBadgeIconView.layer.cornerRadius = WPNotificationBadgeIconSize / 2.0;
+    self.notificationBadgeIconView.layer.masksToBounds = YES;
+    self.notificationBadgeIconView.layer.borderColor = [[UIColor whiteColor] CGColor];
+    self.notificationBadgeIconView.layer.borderWidth = 1.0;
+    self.notificationBadgeIconView.hidden = YES;
+    [self.view addSubview:self.notificationBadgeIconView];
+
+    [self updateNotificationBadgeIconPosition];
+    [self updateNotificationBadgeVisibility];
+}
+
+- (void)updateNotificationBadgeIconPosition
+{
+    UIInterfaceOrientation orientation = [[UIApplication sharedApplication] statusBarOrientation];
+    BOOL isLandscape = (orientation == UIInterfaceOrientationLandscapeLeft) || (orientation == UIInterfaceOrientationLandscapeRight);
+
+    // We need to take the extra space before & after the tabbar into account for iPad and iPhone 6 Plus
+    CGFloat horizontalOffset = 0.0;
+    if (IS_IPAD) {
+        horizontalOffset = isLandscape ? WPNotificationBadgeIconHorizontalOffsetForIPadInLandscape : WPNotificationBadgeIconHorizontalOffsetForIPadInPortrait;
+    }
+    else if (isLandscape && WPDeviceIdentification.isiPhoneSixPlus) {
+        horizontalOffset = WPNotificationBadgeIconHorizontalOffsetForIPhone6PlusInLandscape;
+    }
+    CGFloat verticalPosition = CGRectGetHeight(self.view.bounds) - (self.tabBar.frame.size.height - WPNotificationBadgeIconVerticalOffsetFromTop);
+    // Subtract the space before & after the tabbar
+    CGFloat tabBarContentWidth = self.tabBar.frame.size.width - (horizontalOffset * 2);
+    CGFloat tabItemWidth = tabBarContentWidth / self.tabBar.items.count;
+    CGFloat horizontalPosition = ((WPTabNotifications + 0.5) * tabItemWidth) - WPNotificationBadgeIconHorizontalOffsetFromCenter + horizontalOffset;
+
+    CGRect rect = self.notificationBadgeIconView.frame;
+    rect.origin.x = floorf(horizontalPosition);
+    rect.origin.y = floorf(verticalPosition);
+    self.notificationBadgeIconView.frame = rect;
+}
+
+- (void)animateNotificationBadgeIcon
+{
+    __weak __typeof(self) weakSelf = self;
+    [UIView animateWithDuration:0.3 animations:^{
+        weakSelf.notificationBadgeIconView.transform = CGAffineTransformMakeScale(1.5, 1.5);
+    } completion:^(BOOL finished) {
+        [UIView animateWithDuration:0.3 animations:^{
+            weakSelf.notificationBadgeIconView.transform = CGAffineTransformMakeScale(0.85, 0.85);
+        } completion:^(BOOL finished) {
+            [UIView animateWithDuration:0.2 animations:^{
```

Comment:
Aside from the view being dismissed, the completion block is executed immediately, with `finished == NO` when this code is called:

```objective-c
[view.layer removeAllAnimations];
```

Meaning there's an explicit intention to make the animations stop.  This is why it's important for all animation code to check that parameter and stop chained animations if it's set to `NO`.